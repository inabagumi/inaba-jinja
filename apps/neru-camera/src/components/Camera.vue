<template>
  <div class="camera">
    <div v-if="hasSupportedMediaDevices" class="camera__content">
      <Renderer
        v-if="cameraStream && overlayBlob"
        :camera-stream="cameraStream"
        :overlay-blob="overlayBlob"
        ref="renderer"
      />
    </div>
    <div v-else class="camera__error">
      <h1 class="camera__error__title">カメラの取得に失敗しました</h1>

      <div class="camera__error__body">
        <p class="camera__error__paragraph">
          あなたの使っているアプリはカメラの取得に対応していません。PC
          の場合、Google Chrome と Firefox
          最新版以外での動作は保証できません。Android の場合は Play Store
          からダウンロードできる Chrome を使ってください。
        </p>
        <p class="camera__error__paragraph">
          また iOS では Safari 以外のアプリからカメラの取得ができません。iPhone
          や iPad などの iOS 端末を使っている場合は Safari
          で開き直してください。
        </p>
      </div>
    </div>

    <div class="camera__action-buttons">
      <button
        @click="takePhoto"
        @touchstart="takePhoto"
        class="camera__action-buttons__button"
        :class="{ 'camera__action-buttons__button--active': isShooting }"
        tabindex="-1"
        title="Take a photo!"
      >
        Take a photo!
      </button>
    </div>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import { Component, Prop } from 'vue-property-decorator'

const Renderer = () => import('@/components/Renderer.vue')

const mediaStreamConstraints: MediaStreamConstraints = {
  audio: false,
  video: {
    facingMode: 'environment'
  }
}

type Data = {
  cameraStream?: MediaStream
  isShooting: boolean
  overlayBlob?: Blob
}

type Methods = {
  download: (uri: string) => Promise<Blob>
  takePhoto: () => void
}

type Computed = {
  hasSupportedMediaDevices: boolean
}

type Props = {
  src: string
}

export default Vue.extend<Data, Methods, Computed, Props>({
  components: { Renderer },

  computed: {
    hasSupportedMediaDevices: () => !!navigator.mediaDevices
  },

  data() {
    return {
      cameraStream: undefined,
      isShooting: false,
      overlayBlob: undefined
    }
  },

  methods: {
    async download(uri: string) {
      const response = await fetch(uri)

      return response.blob()
    },

    takePhoto() {
      const renderer = this.$refs.renderer as any

      if (this.isShooting || !renderer) return

      this.isShooting = true

      const anchor = document.createElement('a')

      let otherTab: Window | null
      if (typeof anchor.download !== 'string') {
        otherTab = window.open('about:blank', '_blank')
      }

      renderer
        .toBlob('image')
        .then((blob: Blob | null) => {
          if (!blob) {
            if (otherTab) otherTab.close()

            return
          }

          anchor.href = URL.createObjectURL(blob)

          if (!otherTab) {
            anchor.download = `NeruCamera-${Date.now()}.png`
            anchor.target = '_blank'
            anchor.click()
          } else {
            otherTab.location.href = anchor.href
          }
        })
        .finally(() => {
          this.isShooting = false
        })
    }
  },

  mounted() {
    if (!this.hasSupportedMediaDevices) return

    Promise.all([
      navigator.mediaDevices.getUserMedia(mediaStreamConstraints),
      this.download(this.src)
    ]).then(([cameraStream, overlayBlob]) => {
      this.cameraStream = cameraStream
      this.overlayBlob = overlayBlob
    })
  },

  name: 'Camera',

  props: {
    src: {
      required: true,
      type: String
    }
  }
})
</script>

<style scoped>
.camera {
  align-items: stretch;
  background-color: #000;
  display: flex;
  flex-direction: column;
  height: 100%;
  justify-content: flex-end;
  margin: 0;
}

.camera__content {
  align-items: center;
  display: flex;
  flex-grow: 1;
  overflow: hidden;
}

.camera__action-buttons {
  background-color: #000;
  box-sizing: border-box;
  display: flex;
  justify-content: space-around;
  padding: 5px 12px 24px;
  width: 100%;
}

.camera__action-buttons__button {
  align-items: center;
  background-color: transparent;
  border: 3px solid #fff;
  border-radius: 50%;
  box-sizing: border-box;
  color: #fff;
  display: block;
  height: 56px;
  overflow: hidden;
  padding: 0;
  width: 56px;
}

.camera__action-buttons__button:focus {
  outline: 0;
}

.camera__action-buttons__button::before {
  background-color: #fff;
  border: 2px solid #000;
  border-radius: 50%;
  box-sizing: border-box;
  content: '';
  display: block;
  height: 100%;
  transition: border-width 0.2s linear;
  width: 100%;
}

.camera__action-buttons__button--active::before {
  border-width: 4px;
}

.camera__error {
  padding: 0.5rem;
}

.camera__error__title {
  font-size: 1.25rem;
  font-weight: 700;
  margin: 0 0 0.5rem 0;
}

.camera__error__paragraph {
  line-height: 1.75;
  margin: 0;
}

.camera__error__paragraph:not(:first-child) {
  margin-top: 1rem;
}
</style>
