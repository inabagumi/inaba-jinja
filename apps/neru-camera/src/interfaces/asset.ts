type Asset = {
  id: number
  keyColor: string
  src: string
}

export default Asset
