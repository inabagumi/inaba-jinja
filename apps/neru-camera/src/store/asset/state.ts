export type Asset = {
  id: number
  keyColor: string
  src: string
}

export type AssetState = {
  assets: Asset[]
  isLoading: boolean
}

const state = (): AssetState => ({
  assets: [],
  isLoading: false
})

export default state
