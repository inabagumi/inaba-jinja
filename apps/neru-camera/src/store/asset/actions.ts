import { ActionTree } from 'vuex'
import { RootState } from '../state'
import { Asset, AssetState } from './state'
import {
  ASSETS_FETCH_FAILURE,
  ASSETS_FETCH_REQUEST,
  ASSETS_FETCH_SUCCESS
} from './types'

const actions: ActionTree<AssetState, RootState> = {
  async fetch({ commit }): Promise<void> {
    commit(ASSETS_FETCH_REQUEST)

    let assets: Asset[]

    try {
      const response = await fetch('/list.json')
      assets = await response.json()
    } catch (error) {
      commit(ASSETS_FETCH_FAILURE, error)

      return
    }

    commit(ASSETS_FETCH_SUCCESS, assets)
  }
}

export default actions
