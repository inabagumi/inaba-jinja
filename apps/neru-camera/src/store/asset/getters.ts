import { GetterTree } from 'vuex'
import { RootState } from '../state'
import { AssetState } from './state'

const getters: GetterTree<AssetState, RootState> = {
  getAssetById: state => (id: number) =>
    state.assets.find(asset => asset.id === id) || null
}

export default getters
