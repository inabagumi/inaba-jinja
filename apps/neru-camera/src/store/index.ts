import Vue from 'vue'
import Vuex from 'vuex'
import asset from './asset'
import { RootState } from './state'

Vue.use(Vuex)

export default new Vuex.Store<RootState>({
  modules: { asset }
})
