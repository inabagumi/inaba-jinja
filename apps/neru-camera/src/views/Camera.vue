<template>
  <div class="camera">
    <div class="camera__content">
      <Renderer
        v-if="cameraStream && asset"
        :camera-stream="cameraStream"
        :asset="asset"
        ref="renderer"
      />
    </div>

    <div class="camera__action-buttons">
      <button
        @click="takePhoto"
        @touchstart="takePhoto"
        class="camera__action-buttons__button"
        :class="{ 'camera__action-buttons__button--active': isShooting }"
        tabindex="-1"
        title="Take a photo!"
      >
        Take a photo!
      </button>
    </div>

    <div v-if="!isLoading && !asset" class="camera__error">
      <h1 class="camera__error__title">読み込みに失敗しました</h1>

      <div class="camera__error__body">
        <p class="camera__error__paragraph">
          なんからかの理由で読み込みに失敗しました。
        </p>
      </div>
    </div>

    <div v-else-if="!hasSupportedMediaDevices" class="camera__error">
      <h1 class="camera__error__title">カメラの取得に失敗しました</h1>

      <div class="camera__error__body">
        <p class="camera__error__paragraph">
          あなたの使っているアプリはカメラの取得に対応していません。PC
          の場合、Google Chrome と Firefox
          最新版以外での動作は保証できません。Android の場合は Play Store
          からダウンロードできる Chrome を使ってください。
        </p>
        <p class="camera__error__paragraph">
          また iOS では Safari 以外のアプリからカメラの取得ができません。iPhone
          や iPad などの iOS 端末を使っている場合は Safari
          で開き直してください。
        </p>
      </div>
    </div>

    <div v-else-if="hasError" class="camera__error">
      <h1 class="camera__error__title">カメラの取得に失敗しました</h1>

      <div class="camera__error__body">
        <p class="camera__error__paragraph">
          カメラの取得に失敗しました。ねるカメラを利用するにはカメラへのアクセスを許可する必要があります。
        </p>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import { mapGetters, mapState } from 'vuex'
import { Asset } from '@/store/asset/state'

const AsyncRenderer = () =>
  import(
    /* webpackChunkName: 'components/renderer' */ '@/components/Renderer.vue'
  )

const mediaStreamConstraints: MediaStreamConstraints = {
  audio: false,
  video: {
    facingMode: 'environment',
    height: 960,
    width: 1280
  }
}

type Data = {
  cameraStream?: MediaStream
  hasError: boolean
  isShooting: boolean
}

type Methods = {
  takePhoto: () => void
}

type Computed = {
  asset: Asset
  assetId: number
  getAssetById: (id: number) => Asset | null
  hasSupportedMediaDevices: boolean
  isLoading: boolean
}

type Props = {}

export default Vue.extend<Data, Methods, Computed, Props>({
  components: { Renderer: AsyncRenderer },

  computed: {
    ...(mapGetters('asset', ['getAssetById']) as any),
    ...(mapState('asset', ['isLoading']) as any),

    asset() {
      return this.getAssetById(this.assetId)
    },

    assetId() {
      const value = this.$route.params.id
      const id = Array.isArray(value) ? value[0] : value

      return id ? parseInt(id, 10) : 2
    },

    hasSupportedMediaDevices: () => !!navigator.mediaDevices
  },

  data() {
    return {
      cameraStream: undefined,
      hasError: false,
      isShooting: false
    }
  },

  methods: {
    takePhoto() {
      const renderer = this.$refs.renderer

      if (this.isShooting || !renderer) return

      this.isShooting = true

      const anchor = document.createElement('a')

      let otherTab: Window | null
      if (typeof anchor.download !== 'string') {
        otherTab = window.open('about:blank', '_blank')
      }

      ;(renderer as any)
        .toBlob('image/jpeg')
        .then((blob: Blob | null) => {
          if (!blob) {
            if (otherTab) otherTab.close()

            return
          }

          anchor.href = URL.createObjectURL(blob)

          if (!otherTab) {
            anchor.download = `NeruCamera-${Date.now()}.jpg`
            anchor.target = '_blank'
            anchor.click()
          } else {
            otherTab.location.href = anchor.href
          }
        })
        .finally(() => {
          this.isShooting = false
        })
    }
  },

  mounted() {
    if (!this.hasSupportedMediaDevices) return

    navigator.mediaDevices
      .getUserMedia(mediaStreamConstraints)
      .then(cameraStream => (this.cameraStream = cameraStream))
      .catch(() => (this.hasError = true))
  },

  name: 'Camera'
})
</script>

<style scoped>
.camera {
  align-items: stretch;
  display: flex;
  flex-direction: column;
  height: 100%;
  justify-content: flex-end;
  margin: 0;
}

.camera__content {
  align-items: center;
  display: flex;
  flex-grow: 1;
  overflow: hidden;
}

.camera__action-buttons {
  box-sizing: border-box;
  display: flex;
  justify-content: space-around;
  padding: 5px 12px 24px;
  width: 100%;
}

.camera__action-buttons__button {
  align-items: center;
  background-color: transparent;
  border: 3px solid #fafafa;
  border-radius: 50%;
  box-sizing: border-box;
  color: #fafafa;
  display: block;
  height: 56px;
  overflow: hidden;
  padding: 0;
  width: 56px;
}

.camera__action-buttons__button:focus {
  outline: 0;
}

.camera__action-buttons__button::before {
  background-color: #fafafa;
  border: 2px solid #1b1b1b;
  border-radius: 50%;
  box-sizing: border-box;
  content: '';
  display: block;
  height: 100%;
  transition: border-width 0.2s linear;
  width: 100%;
}

.camera__action-buttons__button--active::before {
  border-width: 4px;
}

.camera__error {
  background-color: rgba(0, 0, 0, 0.8);
  bottom: 0;
  color: #fafafa;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  left: 0;
  padding: 3rem 1rem 0.5rem;
  position: fixed;
  right: 0;
  top: 0;
}

.camera__error__title {
  font-size: 1.25rem;
  font-weight: 700;
  margin: 0 0 0.5rem 0;
}

.camera__error__paragraph {
  line-height: 2;
  margin: 0;
  text-align: justify;
}

.camera__error__paragraph:not(:first-child) {
  margin-top: 1rem;
}
</style>
