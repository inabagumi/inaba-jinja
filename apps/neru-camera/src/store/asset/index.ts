import { Module } from 'vuex'
import { RootState } from '../state'
import actions from './actions'
import getters from './getters'
import mutations from './mutations'
import state, { AssetState } from './state'

const mod: Module<AssetState, RootState> = {
  actions,
  getters,
  mutations,
  namespaced: true,
  state
}

export default mod
