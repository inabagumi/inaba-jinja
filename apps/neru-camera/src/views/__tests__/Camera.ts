import { createLocalVue, shallowMount } from '@vue/test-utils'
import Vuex, { Store } from 'vuex'
import asset from '../../store/asset'
import {
  ASSETS_FETCH_REQUEST,
  ASSETS_FETCH_SUCCESS
} from '../../store/asset/types'
import { RootState } from '../../store/state'
import Camera from '../Camera.vue'

const localVue = createLocalVue()

localVue.use(Vuex)

describe('Camera.vue', (): void => {
  let store: Store<RootState>

  beforeEach((): void => {
    store = new Store({
      modules: { asset }
    })
  })

  it('renders when passed', (): void => {
    store.commit(`asset/${ASSETS_FETCH_REQUEST}`)
    store.commit(`asset/${ASSETS_FETCH_SUCCESS}`, [
      {
        id: 1,
        keyColor: 0x00ff00,
        src: 'https://neru-camera.test/overlay.mp4'
      }
    ])

    const wrapper = shallowMount(Camera, {
      localVue,
      mocks: {
        $route: {
          params: {
            id: 2
          }
        }
      },
      store
    })

    expect(wrapper.text()).toMatchSnapshot()
  })
})
