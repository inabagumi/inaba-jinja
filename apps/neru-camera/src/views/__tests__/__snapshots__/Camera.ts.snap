// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Camera.vue renders when passed 1`] = `
"Take a photo!
     読み込みに失敗しました 
        なんからかの理由で読み込みに失敗しました。"
`;
